KUDO ignores reconcile request for Deletion of StatefulSet

**What happened**:
I tried to write an upgrade plan for an operator.
The upgrade Plan has a step to delete the StatefulSet, and another step to recreate the StatefulSet with different settings.

The delete step COMPLETED successfully, and the recreate step started. This step failed a couple times with
```
Step node [ERROR] (A transient error when executing task upgrade.reinstall.node.node. Will retry. failed to apply StrategicMergePatch to object sanity-test-namespace/cassandra-instance-node: StatefulSet.apps "cassandra-instance-node" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden)
```
This is expected, as the pods of the old StatefulSet are deleted one by one and the StatefulSet can't be updated. After the final Delete of the StatefulSet, no reconcile request came through to trigger a retry of the recreate step.

After update of another resource (a service), the retry worked and the step was executed

**What you expected to happen**:

The deletion of the StatefulSet should trigger a reconcile.

--- a/pkg/controller/instance/instance_controller.go
+++ b/pkg/controller/instance/instance_controller.go
@@ -95,20 +95,6 @@ func (r *Reconciler) SetupWithManager(
 			return requests
 		})
 
-	// resPredicate ignores DeleteEvents for pipe-pods only (marked with task.PipePodAnnotation). This is due to an
-	// inherent race that was described in detail in #1116 (https://github.com/kudobuilder/kudo/issues/1116)
-	// tl;dr: pipe-task will delete the pipe pod at the end of the execution. this would normally trigger another
-	// Instance reconciliation which might end up copying pipe files twice. we avoid this by explicitly ignoring
-	// DeleteEvents for pipe-pods.
-	resPredicate := predicate.Funcs{
-		CreateFunc: func(event.CreateEvent) bool { return true },
-		DeleteFunc: func(e event.DeleteEvent) bool {
-			return !isForPipePod(e)
-		},
-		UpdateFunc:  func(event.UpdateEvent) bool { return true },
-		GenericFunc: func(event.GenericEvent) bool { return true },
-	}
-
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&kudov1beta1.Instance{}).
 		Owns(&kudov1beta1.Instance{}).
@@ -117,9 +103,25 @@ func (r *Reconciler) SetupWithManager(
 		Owns(&batchv1.Job{}).
 		Owns(&appsv1.StatefulSet{}).
 		Owns(&corev1.Pod{}).
-		WithEventFilter(resPredicate).
+		WithEventFilter(eventFilter()).
 		Watches(&source.Kind{Type: &kudov1beta1.OperatorVersion{}}, &handler.EnqueueRequestsFromMapFunc{ToRequests: addOvRelatedInstancesToReconcile}).
 		Complete(r)
+}
+
+// eventFilter ignores DeleteEvents for pipe-pods only (marked with task.PipePodAnnotation). This is due to an
+// inherent race that was described in detail in #1116 (https://github.com/kudobuilder/kudo/issues/1116)
+// tl;dr: pipe-task will delete the pipe pod at the end of the execution. this would normally trigger another
+// Instance reconciliation which might end up copying pipe files twice. we avoid this by explicitly ignoring
+// DeleteEvents for pipe-pods.
+func eventFilter() predicate.Funcs {
+	return predicate.Funcs{
+		CreateFunc: func(event.CreateEvent) bool { return true },
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return !isForPipePod(e)
+		},
+		UpdateFunc:  func(event.UpdateEvent) bool { return true },
+		GenericFunc: func(event.GenericEvent) bool { return true },
+	}
 }
 
 func isForPipePod(e event.DeleteEvent) bool {

--- a/pkg/controller/instance/instance_controller.go
+++ b/pkg/controller/instance/instance_controller.go
@@ -103,8 +103,7 @@ func (r *Reconciler) SetupWithManager(
 	resPredicate := predicate.Funcs{
 		CreateFunc: func(event.CreateEvent) bool { return true },
 		DeleteFunc: func(e event.DeleteEvent) bool {
-			return e.Meta.GetAnnotations() != nil &&
-				funk.Contains(e.Meta.GetAnnotations(), task.PipePodAnnotation)
+			return !isForPipePod(e)
 		},
 		UpdateFunc:  func(event.UpdateEvent) bool { return true },
 		GenericFunc: func(event.GenericEvent) bool { return true },
@@ -121,6 +120,10 @@ func (r *Reconciler) SetupWithManager(
 		WithEventFilter(resPredicate).
 		Watches(&source.Kind{Type: &kudov1beta1.OperatorVersion{}}, &handler.EnqueueRequestsFromMapFunc{ToRequests: addOvRelatedInstancesToReconcile}).
 		Complete(r)
+}
+
+func isForPipePod(e event.DeleteEvent) bool {
+	return e.Meta.GetAnnotations() != nil && funk.Contains(e.Meta.GetAnnotations(), task.PipePodAnnotation)
 }
 
 // Reconcile is the main controller method that gets called every time something about the instance changes

--- a/pkg/controller/instance/instance_controller_test.go
+++ b/pkg/controller/instance/instance_controller_test.go
@@ -12,6 +12,9 @@ import (
 	"testing"
 	"time"
 
+	"github.com/kudobuilder/kudo/pkg/engine/task"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
 	"github.com/stretchr/testify/assert"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -346,5 +349,34 @@ func TestSpecParameterDifference(t *testing.T) {
 	for _, test := range testParams {
 		diff := v1beta1.ParameterDiff(old, test.new)
 		assert.Equal(t, test.diff, diff)
+	}
+}
+
+func TestEventFilterForDelete(t *testing.T) {
+	var testParams = []struct {
+		name    string
+		allowed bool
+		e       event.DeleteEvent
+	}{
+		{"A Pod without annotations", true, event.DeleteEvent{
+			Meta:               &v1.Pod{},
+			Object:             nil,
+			DeleteStateUnknown: false,
+		}},
+		{"A Pod with pipePod annotation", false, event.DeleteEvent{
+			Meta: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{task.PipePodAnnotation: "true"},
+				},
+			},
+			Object:             nil,
+			DeleteStateUnknown: false,
+		}},
+	}
+
+	filter := eventFilter()
+	for _, test := range testParams {
+		diff := filter.Delete(test.e)
+		assert.Equal(t, test.allowed, diff, test.name)
 	}
 }

